src/phoenix/server/api/subscriptions.py

```diff
@@ -63,42 +97,309 @@ async def chat_completion(
+                        name=PLAYGROUND_PROJECT_NAME,
+                        description="Traces from prompt playground",
+                    )
+                )
+            db_span = span.add_to_session(session, playground_project_id)
+            await session.flush()
+            yield FinishedChatCompletion(span=to_gql_span(db_span))
+        info.context.event_queue.put(SpanInsertEvent(ids=(playground_project_id,)))
+
+    @strawberry.subscription
+    async def chat_completion_over_dataset(
+        self, info: Info[Context, None], input: ChatCompletionOverDatasetInput
+    ) -> AsyncIterator[ChatCompletionSubscriptionPayload]:
+        provider_key = input.model.provider_key
+        llm_client_class = PLAYGROUND_CLIENT_REGISTRY.get_client(provider_key, input.model.name)
+        if llm_client_class is None:
+            raise BadRequest(f"No LLM client registered for provider '{provider_key}'")
+
+        dataset_id = from_global_id_with_expected_type(input.dataset_id, Dataset.__name__)
+        version_id = (
+            from_global_id_with_expected_type(
+                global_id=input.dataset_version_id, expected_type_name=DatasetVersion.__name__
+            )
+            if input.dataset_version_id
+            else None
+        )
+        revision_ids = (
+            select(func.max(models.DatasetExampleRevision.id))
+            .join(models.DatasetExample)
+            .where(models.DatasetExample.dataset_id == dataset_id)
+            .group_by(models.DatasetExampleRevision.dataset_example_id)
+        )
+        if version_id:
+            version_id_subquery = (
+                select(models.DatasetVersion.id)
+                .where(models.DatasetVersion.dataset_id == dataset_id)
+                .where(models.DatasetVersion.id == version_id)
+                .scalar_subquery()
+            )
+            revision_ids = revision_ids.where(
+                models.DatasetExampleRevision.dataset_version_id <= version_id_subquery
+            )
+        query = (
+            select(models.DatasetExampleRevision)
+            .where(
+                and_(
+                    models.DatasetExampleRevision.id.in_(revision_ids),
+                    models.DatasetExampleRevision.revision_kind != "DELETE",
+                )
+            )
+            .order_by(models.DatasetExampleRevision.dataset_example_id.asc())
+            .options(
+                load_only(
+                    models.DatasetExampleRevision.dataset_example_id,
+                    models.DatasetExampleRevision.input,
+                )
+            )
+        )
+        async with info.context.db() as session:
+            revisions = [revision async for revision in await session.stream_scalars(query)]
+        if not revisions:
+            raise BadRequest("No examples found for the given dataset and version")
+
+        spans: dict[DatasetExampleID, streaming_llm_span] = {}
+        async for payload in _merge_iterators(
```

Comment:
do we want an `asyncio.gather` here or something?

Reply:
I'm not clear how that would work, let's sync.

Reply:
resolved offline. the behavior of `_merge_iterators` is different because we are handling async streams and need to yield each payload as soon as possible to get a streaming experience in the ui